### ISSUE TITLE
Kodi Frequently Drops Internet Audio Streams With libcurl >= 7.76.0

### ISSUE THREAD

**Issue by stobb3s** [xbmc/xbmc#19838]
### Describe the bug

I use Kodi to listen to audio podcasts on a daily basis using the RSS Podcasts plugin (available in official Matrix add-on repo). Starting several weeks ago, I noticed that some podcasts would stop playing in the middle of an episode, seemingly randomly. The first few times this happened, I assumed it was some type of network quality issue. But the problem continued, and it seemed to only affect certain podcasts. In general, I have observed that podcasts with longer episodes are more likely to drop.

## Expected Behavior

Kodi does not stop playback in the middle of an episode.

## Actual Behavior

Kodi occasionally stops playback in the middle of an episode, depending on some unknown factors.

## Possible Fix

I did some investigation to determine the root cause of the issue. First, I took a [log](https://paste.kodi.tv/basosumepo.kodi). The log was taken with Kodi in Debug Mode and component-specific logging for libcurl enabled. Here is a summary of the log:

1. At 07:39:55 local time, playback begins
2. At 07:57:46 (18 minutes into playback), there are several problems in Curl::Debug. Notably: the message "HTTP/2 stream 1 was not closed cleanly before end of the underlying stream"
3. Curl::Debug reports an ERROR "Failed: Stream error in the HTTP/2 framing layer(92)"
4. CFileCache::Process starts reporting warnings. First "source read returned -1! Will retry", then "source read returned 0! Will retry", which continues for the next 22 minutes
5. At 08:19:29, CFileCache::Process reports an ERROR: "source read didn't return any data before eof!". At this point, playback stops, 40 minutes into a 3 hour and 20 minute episode.

This problem started about the same time as I upgraded to Kodi version 19.1, so I suspected that some recent commits to Kodi were causing the problem. But after searching and testing, I was not able to find anything. I tried downgrading Kodi to version 19.0, but the problem remained.

Then, I looked at recent changes to curl. In the [changelog for version 7.76.0](https://curl.se/changes.html#7_76_0), I noticed there was a bugfix titled "[http2: fail if connection terminated without END_STREAM](https://github.com/curl/curl/pull/6736)", which links to a GitHub PR. In the conversation, they discuss that they are changing the behavior of curl to return an error if the stream is not "correctly closed". In the commit, the code is changed so that libcurl will return an error if the read buffer is 0 bytes length AND the stream is not closed. The error message for that event is "HTTP/2 stream %d was not closed cleanly before end of the underlying stream", which is identical to the error message in my Kodi log.

After this discovery, I downgraded my curl package to version 7.75.0, which fixed the problem. Therefore, I can conclude with confidence that this recent commit to curl is the source of this bug.

I am not sure how to proceed with this finding. Does Kodi need to change to handle the new functionality of curl? Or is the new functionality of curl wrong and I should report it to the curl devs?

### To Reproduce

1. Obtain a Linux build of Matrix that links to libcurl 7.76.0 or later.
2. Instead of downloading and configuring the RSS Podcasts plugin, you may simply put [this URL](https://mp3s.nashownotes.com/NA-1349-2021-05-23-Final.mp3) in a .m3u playlist file and open the stream. It should reproduce the bug just the same.
3. The playback should stop around 45 minutes or so into playback. I think it's dependent on some specific network conditions. The playback always stops for me, the position is just slightly different each time.

### Debuglog

The debuglog can be found here: [Log (same as the one linked above)](https://paste.kodi.tv/basosumepo.kodi)

### Screenshots 
Here are some links or screenshots to help explain the problem: N/A

## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help: N/A

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [ ] tvOS
 - [x] Linux
 - [ ] OSX
 - [ ] Windows
 - [ ] Windows UWP

 - Operating system version/name: Arch Linux, updated about a week ago
 - Kodi version: 19.1, compiled from the official release source code tarball

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

**Comment by fuzzard** [comment]
I'll leave this open for now, but i think you are going to find linux builds are dependant on your system libs, ie we dont ship/build linux statically with libs. So as you found out, the fix is to "fix"/update your system libs.

Good investigative work though, and will help us avoid a potential issue for other platforms when they have their curl static libs bumped

**Comment by sopparus** [comment]
As a work around you can disable http2 in advancedsettings. Xml

**Comment by fuzzard** [comment]
Once again, on linux we dont ship/build with static libs. if a user chooses to do so, go ahead, but theres nothing we can do about that user choosing to link (or use a system lib) of a particular version on linux. My comment about static libs is purely about other platforms (Apple, Windows, Android) that actively use static libs, of which none currently use the version you mention.

Its really nothing wrong with kodi, its as you pointed out an issue with libcurl, so thanks for the info, hopefully others may find it handy if they come across a similar issue, but theres no point in leaving this issue open on our end

**Comment by arnova** [comment]
> I disagree, this seems to be related purely to the changed behavior in curl 7.76.0 and has nothing to do with static/dynamic linking.
> I guess we'll have to look into it, since sooner rather then later this will pop up in other distros and when we bump curl for depends.
> 
> @arnova what's your opinion?

Yes I agree with @wsnipex: Sooner or later this will become a problem for our "standard" builds too, so we better look into it now. However, looking at the amount of http2 problems we're seeing I'm afraid there are simply still a lot of bugs/incompatibilities in libcurl's implementation so there may not much we can do.

@wsnipex: Perhaps we should even consider disabling http2 by default in stable releases for now?
